### PR TITLE
NIET DEZE MERGEN

### PR DESCRIPTION
**Pull request by jim3101**

### CODE REVIEW

File: dpgan.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    gen_optimizer.step()
+
+    return gen_loss, generated_sentence
+
+
+def pretrain_discriminator(input_tensor, target_tensor, generated_sentence, discriminator, disc_optimizer):
+    disc_optimizer.zero_grad()
+    
+    # Determine the discriminator loss for both the true sample and the generated sample
+    generated_disc_loss, generated_disc_out = discriminator(input_tensor, generated_sentence, true_sample=False)
```
Comment (jorism1993) [inline dpgan.py]: Moet dit niet generated_sentece.detach() zijn?

File: dpgan.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    return disc_loss
+
+
+def train(input_tensor, target_tensor, generator, discriminator, adverserial_loss, gen_optimizer, disc_optimizer, train_generator):
+    if train_generator:
+        gen_optimizer.zero_grad()
+    disc_optimizer.zero_grad()
+
+    # Generate a sentence
+    gen_loss, generated_sentence = generator(input_tensor, target_tensor)
```
Comment (jorism1993) [inline dpgan.py]: Hier moet ergens een detach(), denk ik

File: models/Discriminator.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		encoder_input = context_tensor
+		encoder_input_length = encoder_input.shape[1]
+
+		encoder_hidden = self.encoder.initHidden(batch_size=batch_size)
+		encoder_outputs = torch.zeros(encoder_input_length, batch_size, self.encoder.hidden_size, device=DEVICE)
+
+		# Perform the encoder steps for each word in the input tensor
+		for ei in range(encoder_input_length):
+			encoder_output, encoder_hidden = self.encoder(encoder_input[:, ei], encoder_hidden)
+			encoder_outputs[ei, :, :] = encoder_output[0, :, :]
```
Comment (jorism1993) [inline models/Discriminator.py]: encoder_outputs[ei + MAX_LENGTH - encoder_input_length, :, :] werkt iets beter
